### ISSUE TITLE
SIGQUIT causing thread dumps for jvm running inside docker container

### ISSUE THREAD

**Issue by pratiksharkin** [DataDog/jmxfetch#389]
hello @remeh 

We run Datadog agent inside our docker container for metric collection purpose.
We see that JMXFetch is spawned by datadog process like below:

`root     43334 43252  1 08:50 ?        00:03:35 java -Xmx200m -Xms50m -classpath /opt/datadog-agent/bin/agent/dist/jmx/jmxfetch.jar org.datadog.jmxfetch.App --ipc_host localhost --ipc_port 5001 --check_period 15000 --thread_pool_size 3 --collection_timeout 60 --reconnection_timeout 10 --reconnection_thread_pool_size 3 --log_level INFO --reporter statsd:localhost:8125 collect`

One of our java process looks something like this:

`ubuntu   18067 17637  0 Mar08 ?        00:04:37 java -XX:+UseParallelGC -XX:NativeMemoryTracking=summary -Xss256K -Dorg.xerial.snappy.tempdir=/home/ubuntu/tmp -Dlog4j2.debug=DEBUG -Dlogdir=/var/log/abclog -Dlog4j.configurationFile=/home/ubuntu/build-target/latency-metrics/latency-metrics.log4j2.xml -Dlog4j2.formatMsgNoLookups=true -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.password.file=/home/ubuntu/build-target/latency-metrics/jmxremote.password -Dcom.sun.management.jmxremote.access.file=/home/ubuntu/build-target/latency-metrics/jmxremote.access -Djava.rmi.server.hostname=10.79.199.11 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.port=11110 -Dcom.sun.management.jmxremote.password.toHashes=false -Xmx128m -Dreporters.configuration=/home/ubuntu/build-target/latency-metrics/reporters.configuration -Dreporters.application.name=App -Ddisable.task.reporters=true -Ddatadog.host=10.19.129.111 -Ddatadog.tags=env:APP-NEW-DEV -DApp.metrics.host=machine.10.79.199.11 -Dpostgres.configuration=/home/ubuntu/build-target/latency-metrics/postgres.configuration -Ddynamodb.configuration=/home/ubuntu/build-target/latency-metrics/dynamodb.configuration -Delasticsearch.configuration=/home/ubuntu/build-target/latency-metrics/elasticsearch.configuration -Ddpconfig.base.folder=/home/ubuntu/build-target/latency-metrics -Ddp.operational.config.base.folder=/home/ubuntu/build-target/latency-metrics -Dtaskmgr.spec.folder=/home/ubuntu/build-target/latency-metrics -DOvaParamFile=/etc/app/ova.params -Dcustomer.configuration=/home/ubuntu/build-target/latency-metrics/customer.configuration -Dpolicy.config.path=/home/ubuntu/build-target/latency-metrics/policy -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/var/lib/heap-dumps/latency-metrics -XX:+ExitOnOutOfMemoryError --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio=ALL-UNNAMED --add-exports java.base/jdk.internal.misc=ALL-UNNAMED --add-modules jdk.unsupported -Dio.netty.tryReflectionSetAccessible=true --illegal-access=warn -Djava.security.properties==/home/ubuntu/build-target/latency-metrics/java.security.fips -Dorg.bouncycastle.jca.enable_jks=true -cp /home/ubuntu/build-target/latency-metrics/App-0.001-SNAPSHOT.jar com.app.latency.Main`

While JMXFetch is running, we continously see thread dumps getting printed inside our Java process's log file.

in the /var/log/datadog/jmxfetch.log we see following:
`2022-03-09 12:06:21 UTC | JMX | DEBUG | Reporter | Instance TestApp-11108 is sending 27 metrics to the metrics reporter during collection #858
2022-03-09 12:06:21 UTC | JMX | DEBUG | Reporter | Instance MyApp-11103 is sending 27 metrics to the metrics reporter during collection #858`

When we kill the datadog-agent (which kills JMXFetch too) the thread dumps stop.
Any idea whats causing this behaviour ?

**Comment by remeh** [comment]
Hi @pratiksharkin,
JMXFetch is using the JMX interface that your app is exposing to collect beans information (i.e. metrics), my guess here would be that your test app is using a thread on each collection call and is logging a thread dump. Would that make sense? Any chance you can share a sample of the log?

**Comment by pratiksharkin** [comment]
@remeh 
Our App exposes a port on which JMX requests can be sent. Possibly a thread is spawned (like Attach Listener), but we haven't written its implementation. 
Also, This behaviour is not observed when we run the same process on a vm. This is observed only inside docker container.

The log dump looks like following:

`2022-03-09 11:53:15
Full thread dump OpenJDK 64-Bit Server VM (11.0.13+8-LTS mixed mode):

Threads class SMR info:
_java_thread_list=0x00007fe12f6123c0, length=72, elements={
0x00007fe16ecb3000, 0x00007fe16ecb4000, 0x00007fe16ecb5000, 0x00007fe16ecb6000,
0x00007fe16ecb6800, 0x00007fe16ecb7800, 0x00007fe16ecb8800, 0x00007fe16ecbe000,
0x00007fe16ecbf000, 0x00007fe16ecc0000, 0x00007fe16ecc1000, 0x00007fe16ecc3800,
0x00007fe1281c4000, 0x00007fe1281c3000, 0x00007fe1281c5800, 0x00007fe1281c6800,
0x00007fe129a1d000, 0x00007fe129a1e000, 0x00007fe1289ee000, 0x00007fe1289ee800,
0x00007fe1538c4000, 0x00007fe1312ae000, 0x00007fe16ecb2800, 0x00007fe12854b000,
0x00007fe130713000, 0x00007fe130a26000, 0x00007fe12c054800, 0x00007fe12b4c2800,
0x00007fe12c055000, 0x00007fe12ba32000, 0x00007fe12c642000, 0x00007fe12c643000,
0x00007fe131690800, 0x00007fe129a1c000, 0x00007fe128550800, 0x00007fe1312ac800
}

"Reference Handler" #2 daemon prio=10 os_prio=0 cpu=21.26ms elapsed=20519.53s tid=0x00007fe16ecb3000 nid=0x2b78 waiting on condition  [0x00007fe1729dd000]
   java.lang.Thread.State: RUNNABLE
        at java.lang.ref.Reference.waitForReferencePendingList(java.base@11.0.13/Native Method)
        at java.lang.ref.Reference.processPendingReferences(java.base@11.0.13/Reference.java:241)
        at java.lang.ref.Reference$ReferenceHandler.run(java.base@11.0.13/Reference.java:213)
        .....
        .....
        .....
`
What metrics does JMXFetch collects ? When we tried other commands like `jcmd VM.native_memory summary`, `jmap -histo:live` We don't see similar thread dumps.

**Comment by pratiksharkin** [comment]
@remeh 
I tried the jmxterm tool, and got connected to our java application:
I tried reading  mbean  - java.lang=Threading :

```
$>domain java.lang
#domain is set to java.lang
$>bean java.lang:type=Threading
#bean is set to java.lang:type=Threading
$>info
#mbean = java.lang:type=Threading
#class name = com.sun.management.internal.HotSpotThreadImpl
# attributes
  %0   - AllThreadIds ([J, r)
  %1   - CurrentThreadAllocatedBytes (long, r)
  %2   - CurrentThreadCpuTime (long, r)
  %3   - CurrentThreadCpuTimeSupported (boolean, r)
  %4   - CurrentThreadUserTime (long, r)
  %5   - DaemonThreadCount (int, r)
  %6   - ObjectMonitorUsageSupported (boolean, r)
  %7   - ObjectName (javax.management.ObjectName, r)
  %8   - PeakThreadCount (int, r)
  %9   - SynchronizerUsageSupported (boolean, r)
  %10  - ThreadAllocatedMemoryEnabled (boolean, rw)
  %11  - ThreadAllocatedMemorySupported (boolean, r)
  %12  - ThreadContentionMonitoringEnabled (boolean, rw)
  %13  - ThreadContentionMonitoringSupported (boolean, r)
  %14  - ThreadCount (int, r)
  %15  - ThreadCpuTimeEnabled (boolean, rw)
  %16  - ThreadCpuTimeSupported (boolean, r)
  %17  - TotalStartedThreadCount (long, r)
# operations
  %0   - [Ljavax.management.openmbean.CompositeData; dumpAllThreads(boolean p0,boolean p1)
  %1   - [Ljavax.management.openmbean.CompositeData; dumpAllThreads(boolean p0,boolean p1,int p2)
  %2   - [J findDeadlockedThreads()
  %3   - [J findMonitorDeadlockedThreads()
  %4   - [J getThreadAllocatedBytes([J p0)
  %5   - long getThreadAllocatedBytes(long p0)
  %6   - [J getThreadCpuTime([J p0)
  %7   - long getThreadCpuTime(long p0)
  %8   - javax.management.openmbean.CompositeData getThreadInfo(long p0,int p1)
  %9   - javax.management.openmbean.CompositeData getThreadInfo(long p0)
  %10  - [Ljavax.management.openmbean.CompositeData; getThreadInfo([J p0,int p1)
  %11  - [Ljavax.management.openmbean.CompositeData; getThreadInfo([J p0,boolean p1,boolean p2,int p3)
  %12  - [Ljavax.management.openmbean.CompositeData; getThreadInfo([J p0,boolean p1,boolean p2)
  %13  - [Ljavax.management.openmbean.CompositeData; getThreadInfo([J p0)
  %14  - [J getThreadUserTime([J p0)
  %15  - long getThreadUserTime(long p0)
  %16  - void resetPeakThreadCount()
#there's no notifications
$>get ThreadCount
#mbean = java.lang:type=Threading:
ThreadCount = 67;

$>
$>get PeakThreadCount
#mbean = java.lang:type=Threading:
PeakThreadCount = 72;

```

While executing these commands I did not see the similar thread dumps in applications log.
Can you tell us which mbeans does JMXFetch must be collecting ? I can try them out through jmxterm
and see if anyone of them reproduces our issue.

**Comment by remeh** [comment]
Interesting, can you also try running the `beans` command with `JMXTerm`? I would have expected a thread dumps output within your app.

In order to know what JMXFetch would be collecting, you can run the command:

```
sudo -u dd-agent datadog-agent jmx list everything
```

(note that in order to run a command with the Datadog Agent binary, your mileage may vary depending on how you deploy it, but the idea is to run `agent jmx list everything`)

What version of the Agent are you running?

**Comment by pratiksharkin** [comment]
@remeh We were able to RCA our issue.

Essentially, one of our metric collection script (which is called by jmxfetch) was calling 'jcmd' as root user.
We run our container in an unprivileged mode, the socket file used for jmx communication is only accessible to user which
has spawned the given jvm process (against which you are running jcmd). Not even root can access it.
The file's location is : /proc/<pid>/root/tmp/.javapid<pid> 
we were able to fix this by calling jcmd as same user as that of java process.
Thank you for your help !
